src/immer.d.ts

```diff
@@ -92,32 +67,32 @@ export interface IProduce {
      * @param {Function} patchListener - optional function that will be called with all the patches produced here
      * @returns {any} a new state, or the initial state if nothing was modified
      */
-    <Base = any, Return = void>(
-        base: Base extends Function ? never : Base,
-        recipe: (this: Draft<Base>, draft: Draft<Base>) => Return,
+
+    /** Curried producer */
+    <
+        Recipe extends (...args: any[]) => any,
+        Params extends any[] = Parameters<Recipe>,
+        T = Params[0],
+    >(
+        recipe: Recipe,
+    ): (state: Immutable<T>, ...rest: Tail<Params>) => Produced<Immutable<T>, ReturnType<Recipe>>
+
+    /** Curried producer with initial state */
+    <
+        Recipe extends (...args: any[]) => any,
+        Params extends any[] = Parameters<Recipe>,
+        T = Params[0],
+    >(
+        recipe: Recipe,
+        initialState: T
+    ): (state?: Immutable<T>, ...rest: Tail<Params>) => Produced<Immutable<T>, ReturnType<Recipe>>
+
+    /** Normal producer */
+    <Base, D = Draft<Base>, Return = void>(
```

Comment:
Separating Base and D allows things like: `produce<readonly string[]>([], (draft: string[]) => ...`